### ISSUE TITLE
[proposal] [util] torch.to(obj, device) supporting reecursive lists/dicts/tuples of tensors

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#69431]
Often it is needed to move model results to cpu (or inputs to gpu). Once the data structures get a bit complicated, dicts and lists appear often in model results. Often we have to roll a little utility method like below. If indeed other people had to write this sort of utilities, may make sense to include something like this to core.

```python
def to(obj, device):
  if torch.is_tensor(obj):
    return obj.to(device)
  if isinstance(obj, dict):
    return {k : to(v, device) for k, v in obj.items()}
  if isinstance(obj, tuple):
    return tuple(to(v, device) for v in obj)
  if isinstance(obj, list):
    return [to(v, device) for v in obj]
  return obj
```

**Comment by albanD** [comment]
Wouldn't this be nicely solved by just using pytree from [here](https://github.com/pytorch/pytorch/blob/3edf1b6ceeaa1e951ed615e9b3b64f0d30c14780/torch/utils/_pytree.py#L177)
where you can do `tree_map(torch.to, your_obj)` ?

**Comment by vadimkantorov** [comment]
maybe! but I think it's still useful to have this primitive in core as it appears a lot in user code and very few people know about pytrees (i certainly don't)

also, torch.to doesn't exist now or isn't documented

**Comment by albanD** [comment]
> also, torch.to doesn't exist now or isn't documented

Ho I used that because you mentioned it above :p But indeed that doesn't even exist.

**Comment by mruberry** [comment]
The lack of `torch.to` isn't actually an impediment because it's easy to make the method a function using a lambda.

We do write "crawlers" like this occasionally. `assert_close` is the most recent example that comes to mind and has one. I think there's also a slightly different version in `common_methods_invocations.py`. I'm inclined to suggest people write their own crawlers or use a Python utility package for functionality like this. There are a lot of vagaries (like what if the "elements" in the containers aren't other containers or tensors?) and the functionality is more about dealing with Python datastructures than it is dealing with tensors. 

NumPy doesn't have any functionality like this, does it?

**Comment by vadimkantorov** [comment]
NumPy doesn't have a concept of device anyway, so it doesn't have this obviously, but the reason isn't relevant IMO.

**Comment by vadimkantorov** [comment]
I saw somewhere more examples of pytree, so I now understand better that pytree traversal indeed can be used for making this recursive device transfer :) But I would still suggest, that it's a good default behavior for torch.to (to be introduced) without forcing users to make their own pytree'd version.

**Comment by vadimkantorov** [comment]
I also wonder if treemap is parallel (akin to _foreach methods). Probably for `torch.to(..., non_blocking=True)` it doesn't matter as much, but still. It also probably could benefit from a single cudaMalloc call (at least for all strided tensors) flattening all tensors in the input

**Comment by vadimkantorov** [comment]
Related PR that actually implements this recursive utility: https://github.com/pytorch/pytorch/pull/77187

**Comment by rohan-varma** [comment]
Found this issue after @vadimkantorov comment on a related PR. Agreed that such a utility would be quite useful and PT-D would then not need such custom logic to move inputs for DDP / FSDP.

@albanD @jbschlosser @mruberry Do we have any new thoughts on this feature and whether this is something that the core team might be able to address?

**Comment by albanD** [comment]
Any reason for not using `pytree` as suggested above in your case?

**Comment by mruberry** [comment]
It's nice when we don't add sugars on top of core operations.

**Comment by albanD** [comment]
Well, for `nn.Module`, that already works as you can call to on the top Module.

**Comment by vadimkantorov** [comment]
there are some stuffs about streams, not sure if tree_map will be able to handle it?

may also be a good idea to support non_blocking and calling custom methods `.to` if they exist in addition to tensors

**Comment by vadimkantorov** [comment]
So now in theory, torch.to could just be implemented in terms of torch.utils._pytree.tree_map.

Compared to vanilla tree_map, torch.to could theoretically do all allocations in one batch, but not sure what could be a good design for it

**Comment by vadimkantorov** [comment]
torch.to could also support a function that minimizes the number of host-device copies as described in this post: https://pytorch.org/blog/optimizing-production-pytorch-performance-with-graph-transformations/#31-combining-input-sparse-features

**Comment by vadimkantorov** [comment]
although in another context (sparse tensors) this post describes an optimization option that torch.to could also do (minimize the number of host->device copies): https://pytorch.org/blog/optimizing-production-pytorch-performance-with-graph-transformations/#31-combining-input-sparse-features

**Comment by albanD** [comment]
I guess pytree would be the way to go for this in 2024 :D
`new = tree_map_only(torch.Tensor, lambda t: t.to(device), old)`

**Comment by vadimkantorov** [comment]
On the workability side, yes it would cut it, but having some builtin option for separate copy thread or efficiently moving a lot of tensors in one allocation might be interesting. Otheriwse, for functional programming side, I think implementing torch.to as tree_map_only is a worthy shortcut for users to have in core and for consistency between functions/methods :)

**Comment by albanD** [comment]
I do not think that changing torch.to() would be good for consistency actually. That would make the behavior significantly different between function and methods.

**Comment by vadimkantorov** [comment]
For basic individual tensors, the behavior would be exactly the same anyway? So for all inputs acceptable for instance .to method, the function behavior torch.to would be the same, right? I'm proposing to keep them strictly identical to all individual tensors (so, both .to and torch.to would continue to exist), and potentially have more knobs/functionality for more complex inputs, like tensor lists or pytree object hierarchies

**Comment by vadimkantorov** [comment]
Btw `_recursive_to` got copy-pasted in https://github.com/pytorch/pytorch/blob/main/torch/distributed/optim/zero_redundancy_optimizer.py#L29

So maybe it is the right moment for it to be promoted at least to `torch._to`?

**Comment by albanD** [comment]
We should remove this code and just use tree_map there instead. It's going to be a lot more reliable...

**Comment by vadimkantorov** [comment]
> just use tree_map there instead

I think it'd be also a great impl for a publicly advertised shortcut :) Otherwise, this duplicate code exists in distributed_utils and in this ZeroOptimizer.py

A more custom op could somehow parallelize copies (if a large tensor list is passed as input) using multiple CUDA threads or allocate a single large contig memory chunk on GPU (and maybe do all this in async fashion if this suits to be able to schedule the copies on user-provided background CUDA stream used for copies)

**Comment by albanD** [comment]
I think that if we need to add one more API for our users to know about, I prefer for them to learn about pytree. It will allow them to solve many of their problems with one line. Compared to a specialized API that will only solve a single problem.

**Comment by albanD** [comment]
In theory, nothing prevents Dynamo from tracing pytree work and doing fancy optimization there :D 

cc @zou3519 for pytree as a public feature

**Comment by vadimkantorov** [comment]
Btw, funnily, `torch.to(x, some_obscure_int)` existed in TorchScript/printous :)

**Comment by vadimkantorov** [comment]
Seems like an attempt to implement something like `torch.to` :)
- https://github.com/pytorch/pytorch/issues/134577

**Comment by vadimkantorov** [comment]
Maybe the idea of frontend `torch.to` could also be extended with `torch.to(..., dtype = )` and also made working with dataclasses and then https://github.com/pytorch/pytorch/issues/153077 could use `torch.to` :)

**Comment by vadimkantorov** [comment]
Hmm, maybe a high-level wrapper like this could somehow solve the problem with launching hundreds of kernels and filling up kernel launch queue:
- https://github.com/pytorch/pytorch/issues/154272

?

**Comment by vadimkantorov** [comment]
Also, existing `to` is super hard to find in search currently: https://docs.pytorch.org/docs/stable/search.html?q=to&check_keywords=yes&area=default#

It would be much better if `torch.to` existed ;)

**Comment by vadimkantorov** [comment]
a related method `torch.clone(...)` seems to already exist: https://docs.pytorch.org/docs/stable/generated/torch.clone.html

(also unfortunately no `out=` argument for `clone`/`to`)
